src/Umbraco.Web.Common/ApplicationBuilder/UmbracoApplicationBuilder.cs

```diff
@@ -78,18 +84,35 @@ public void WithEndpoints(Action<IUmbracoEndpointBuilderContext> configureUmbrac
         }
 
         /// <summary>
-        /// Registers the default required middleware to run Umbraco
+        /// Registers the default required middleware to run Umbraco.
         /// </summary>
-        /// <param name="umbracoApplicationBuilderContext"></param>
         public void RegisterDefaultRequiredMiddleware()
         {
             UseUmbracoCoreMiddleware();
 
             AppBuilder.UseStatusCodePages();
 
-            // Important we handle image manipulations before the static files, otherwise the querystring is just ignored.            
+            // Important we handle image manipulations before the static files, otherwise the querystring is just ignored.
             AppBuilder.UseImageSharp();
+
+            // Get media file provider and request path/URL
+            IFileProvider mediaFileProvider = AppBuilder.ApplicationServices.GetRequiredService<MediaFileManager>().CreateFileProvider();
+            if (mediaFileProvider is not null)
+            {
+                GlobalSettings globalSettings = AppBuilder.ApplicationServices.GetRequiredService<IOptions<GlobalSettings>>().Value;
+                IHostingEnvironment hostingEnvironment = AppBuilder.ApplicationServices.GetService<IHostingEnvironment>();
+                string mediaRequestPath = hostingEnvironment.ToAbsolute(globalSettings.UmbracoMediaPath);
+
+                // Configure custom file provider for media
+                AppBuilder.UseStaticFiles(new StaticFileOptions()
+                {
+                    FileProvider = mediaFileProvider,
+                    RequestPath = mediaRequestPath
+                });
+            }
+
             AppBuilder.UseStaticFiles();
```

Comment:
I noticed we already do the same for `App_Plugins`:

https://github.com/umbraco/Umbraco-CMS/blob/5bfab13dc5a268714aad2426a2b68ab5561a6407/src/Umbraco.Web.Common/Extensions/ApplicationBuilderExtensions.cs#L119-L127

It might make more sense if we update the `WebRootFileProvider` instead, as mentioned in https://github.com/dotnet/aspnetcore/issues/17409. We already have a dependency on Dazinator.Extensions.FileProviders, so we can use the [`PrependBasePathFileProvider`](https://github.com/dazinator/Dazinator.Extensions.FileProviders#prependbasepathfileprovider) to scope the file providers to specific paths:

```c#
env.WebRootFileProvider = new CompositeFileProvider(
    new PrependBasePathFileProvider(Constants.SystemDirectories.AppPlugins, umbracoPluginFileProvider),
    new PrependBasePathFileProvider(mediaRequestPath, mediaFileProvider),
    env.WebRootFileProvider);
```

This has the added benefit of having requests cascade through the file providers, so you can serve files in the same request path from different providers (e.g. media stored in Azure Blob Storage and static assets added to `wwwroot/media/`).

This would also solve the ImageSharp issues with multiple `IImageProviders`, as the default `PhysicalFileSystemProvider` (which is weirdly named, as it just uses `WebRootFileProvider` and the `IFileProvider` abstraction) should be able to fetch the images from all the composited file providers.

Reply:
I've rewritten the code to use the composited file providers and that seems to work perfectly!